In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [3]:
df = pd.read_csv("country_vaccination_stats.csv")
df.head()

country        date  daily_vaccinations   vaccines
0  Argentina  12/29/2020                 NaN  Sputnik V
1  Argentina  12/30/2020             15656.0  Sputnik V
2  Argentina  12/31/2020             15656.0  Sputnik V
3  Argentina    1/1/2021             11070.0  Sputnik V
4  Argentina    1/2/2021              8776.0  Sputnik V

In [8]:
df["daily_vaccinations"].isna().sum() # We have 60 NaN value but dont know which country has it

60

### Before we start analyzing NaN values, lets find out minimum vaccinations value of each country

In [12]:
country_min_vaccinations = df.groupby("country")["daily_vaccinations"].min() 
country_min_vaccinations

country
Argentina                 6483.0
Austria                   3368.0
Bahrain                    943.0
Belgium                      1.0
Brazil                     112.0
Bulgaria                   472.0
Canada                     866.0
Chile                       51.0
China                   187500.0
Costa Rica                 240.0
Croatia                    989.0
Cyprus                     534.0
Czechia                   1822.0
Denmark                   1994.0
Ecuador                    108.0
England                  37625.0
Estonia                    309.0
Finland                    459.0
France                    5000.0
Germany                  19721.0
Gibraltar                  520.0
Greece                     549.0
Hungary                   1339.0
Iceland                     63.0
India                   112150.0
Indonesia                11823.0
Ireland                    550.0
Isle of Man                 90.0
Israel                    6887.0
Italy                      926.0
Ku

In [13]:
len(country_min_vaccinations)

60

In [14]:
country_min_vaccinations.isna().sum()

1

**With that we learned we have 60 country and only one of them has have all NaN values which is Kuwait**<br>
**We will fill Kuwait's values with zeros**

In [20]:
df.loc[df['country'] == 'Kuwait', 'daily_vaccinations'] = df.loc[df['country'] == 'Kuwait', 'daily_vaccinations'].fillna(0)
print(df[df['country'] == 'Kuwait'])

    country        date  daily_vaccinations         vaccines
750  Kuwait  12/28/2020                 0.0  Pfizer/BioNTech


**We filled Kuwait's values with zeros, now we need to fill other countryies value's with their minimum vaccinations values.**<br>
**We must also be careful not to change the value of Kuwait.**

In [21]:
for country, min_vacc in country_min_vaccinations.items():
    if pd.isna(min_vacc): 
        df.loc[df['country'] == country, 'daily_vaccinations'] = df.loc[df['country'] == country, 'daily_vaccinations'].fillna(0)
    else:
        df.loc[df['country'] == country, 'daily_vaccinations'] = df.loc[df['country'] == country, 'daily_vaccinations'].fillna(min_vacc)


In [33]:
df.isna().sum()


country               0
date                  0
daily_vaccinations    0
vaccines              0
dtype: int64

**Lets check our procees**

In [36]:
turkey_vaccinations = df[df['country'] == 'Turkey']['daily_vaccinations']
print("Turkey's daily vaccinations values:")
print(france_vaccinations)

is_nan_present = turkey_vaccinations.isna().any()
print("\n Do we have ant NaN value ? :", is_nan_present)

if is_nan_present:
    print("\n There are some NaN values, try again")
else:
    print("\n All NaN Values filled with the minimum values of Turkey, procees succed")

Turkey's daily vaccinations values:
1352     67192.0
1353    279452.0
1354    308326.0
1355    225634.0
1356    176613.0
1357    166482.0
1358    158439.0
1359    150356.0
1360    120115.0
1361     83651.0
1362     78569.0
1363     76886.0
1364     67192.0
Name: daily_vaccinations, dtype: float64

 Do we have ant NaN value ? : False

 All NaN Values filled with the minimum values of Turkey, procees succed


**I can see the value 67192 in 2 places, one of these values existed and we added the other.**